internal/protocol/connector.go

```diff
@@ -126,56 +128,98 @@ func (c *Connector) connectAttemptAll(ctx context.Context, log logging.Func) (*P
 		return servers[i].Role < servers[j].Role
 	})
 
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	sem := semaphore.NewWeighted(10)
+
+	protocolChan := make(chan *Protocol)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(len(servers))
+
+	go func() {
+		wg.Wait()
+		close(protocolChan)
+	}()
+
 	// Make an attempt for each address until we find the leader.
 	for _, server := range servers {
-		log := func(l logging.Level, format string, a ...interface{}) {
-			format = fmt.Sprintf("server %s: ", server.Address) + format
-			log(l, format, a...)
-		}
+		go func(server NodeInfo, pc chan<- *Protocol) {
+			defer wg.Done()
 
-		ctx, cancel := context.WithTimeout(ctx, c.config.AttemptTimeout)
-		defer cancel()
+			sem.Acquire(ctx, 1)
```

Comment:
This can return an error. We should handle that case, and if you release too many times, it will cause a panic. We should be careful in that scenario.

Reply:
Updated, following the [example](https://pkg.go.dev/golang.org/x/sync/semaphore#example-package-WorkerPool)

Reply:
There is still [the case](https://cs.opensource.google/go/x/sync/+/refs/tags/v0.6.0:semaphore/semaphore.go;drc=43a5402ce75a95522677f77c619865d66b8c57ab;l=109) where the semaphore can be acquired after cancellation. To sacrifice performance but improve readability we could also eliminate (new) lines 156-158

Reply:
We do exactly the same in juju 😉